File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        prefix = suffix = 1
        result = float("-inf")
        for i in range(len(nums)):
            prefix = nums[i] * (prefix or 1)
            suffix = nums[len(nums)-1-i] * (suffix or 1)
            result = max(result, prefix, suffix)
        return result

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        /*
            [1,2,4,6]
            [1,1,2,8]
            [48,24,6,1]

            [48, 24, 12, 8]

        */
        int length = nums.length;
        int[] output = new int[length];
        int[] prefix = new int[length];
        int[] suffix = new int[length];

        prefix[0] = 1;
        suffix[length - 1 ] = 1;

        for (int i = 1; i < length; i++) {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }

        for (int i = length - 2; i >= 0; i--) {
            suffix[i] = suffix[i + 1] * nums[i + 1];
        }

        for (int i = 0; i < length; i++) {
            output[i] = suffix[i] * prefix[i];
        }
        return output;
    }
}  
